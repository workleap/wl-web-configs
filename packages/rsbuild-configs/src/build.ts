import { defineConfig, type DistPathConfig, type HtmlConfig, type Minify, type RsbuildConfig, type RsbuildEntry, type RsbuildPlugins, type SourceMap } from "@rsbuild/core";
import { pluginImageCompress, type PluginImageCompressOptions } from "@rsbuild/plugin-image-compress";
import { pluginReact, type PluginReactOptions } from "@rsbuild/plugin-react";
import { pluginSvgr, type PluginSvgrOptions } from "@rsbuild/plugin-svgr";
import { SwcJsMinimizerRspackPlugin, type Optimization } from "@rspack/core";
import path from "node:path";
import { applyTransformers, type RsbuildConfigTransformer } from "./applyTransformers.ts";

export type OptimizeOption = boolean | "readable";

export type DefineBuildHtmlPluginConfigFunction = (defaultOptions: HtmlConfig) => HtmlConfig;
export type DefineBuildDefineReactPluginConfigFunction = (defaultOptions: PluginReactOptions) => PluginReactOptions;
export type DefineBuildSvgrPluginConfigFunction = (defaultOptions: PluginSvgrOptions) => PluginSvgrOptions;
export type DefineBuildImageCompressPluginConfigFunction = (defaultOptions: PluginImageCompressOptions) => PluginImageCompressOptions;

export interface DefineBuildConfigOptions {
    entry?: RsbuildEntry;
    // Similar to webpack.output.path.
    distPath?: DistPathConfig;
    // Similar to webpack.publicPath.
    assetPrefix?: string;
    plugins?: RsbuildPlugins;
    html?: false | DefineBuildHtmlPluginConfigFunction;
    minify?: Minify;
    optimize?: OptimizeOption;
    sourceMap?: boolean | SourceMap;
    react?: false | DefineBuildDefineReactPluginConfigFunction;
    svgr? : false | DefineBuildSvgrPluginConfigFunction;
    compressImage?: false | DefineBuildImageCompressPluginConfigFunction;
    environmentVariables?: Record<string, unknown>;
    transformers?: RsbuildConfigTransformer[];
    verbose?: boolean;
}

function defaultDefineHtmlPluginConfig(options: HtmlConfig) {
    return options;
}

function defaultDefineReactPluginConfig(options: PluginReactOptions) {
    return options;
}

function defineSvgrPluginConfig(options: PluginSvgrOptions) {
    return options;
}

function defineImageCompressPluginConfig(options: PluginImageCompressOptions) {
    return options;
}

export function getOptimizationConfig(optimize: OptimizeOption): Optimization {
    if (optimize === true) {
        return {
            minimize: true
        };
    } else if (optimize === "readable") {
        return {
            minimize: true,
            minimizer: [
                new SwcJsMinimizerRspackPlugin({
                    minimizerOptions: {
                        mangle: false,
                        compress: {
                            toplevel: true,
                            hoist_props: false
                        }
                    }
                })
            ],
            chunkIds: "named",
            moduleIds: "named",
            mangleExports: false
        };
    }

    // Doesn't turnoff everything but is good enough to help with debugging scenarios.
    return {
        minimize: false,
        chunkIds: "named",
        moduleIds: "named",
        concatenateModules: false,
        mangleExports: false,
        removeAvailableModules: false,
        usedExports: false
    };
}

export function defineBuildConfig(options: DefineBuildConfigOptions = {}) {
    const {
        entry = {
            index: path.resolve("./src/index.tsx")
        },
        distPath = {
            root: path.resolve("./dist")
        },
        assetPrefix = "/",
        plugins = [],
        html = defaultDefineHtmlPluginConfig,
        minify = true,
        optimize = true,
        sourceMap = {
            js: "source-map",
            css: true
        },
        react = defaultDefineReactPluginConfig,
        svgr = defineSvgrPluginConfig,
        compressImage = defineImageCompressPluginConfig,
        // Using an empty object literal as the default value to ensure
        // "process.env" is always available.
        environmentVariables = {},
        transformers = [],
        verbose = false
    } = options;

    const config: RsbuildConfig = {
        mode: "production",
        source: {
            entry,
            // Stringify the environment variables because the plugin does a direct text replacement. Otherwise, "production" would become production
            // after replacement and cause an undefined var error because the production var doesn't exist.
            // For more information, view: https://rsbuild.dev/guide/advanced/env-vars#using-define.
            define: {
                "process.env": Object.keys(environmentVariables).reduce((acc, key) => {
                    acc[key] = JSON.stringify(environmentVariables[key]);

                    return acc;
                }, {} as Record<string, string>)
            }
        },
        output: {
            target: "web",
            distPath,
            cleanDistPath: true,
            assetPrefix,
            filename: {
                html: "[name].html",
                js: "[name].js",
                css: "[name].css",
                svg: "[name].[contenthash:8].svg",
                font: "[name].[contenthash:8][ext]",
                image: "[name].[contenthash:8][ext]",
                media: "[name].[contenthash:8][ext]",
                assets: "[name].[contenthash:8][ext]"
            },
            minify,
            sourceMap
        },
        html: html
            ? html({ template: path.resolve("./public/index.html") })
            : undefined,
        plugins: [
            react && pluginReact(react({
                fastRefresh: false
            })),
            svgr && pluginSvgr(svgr({
                svgrOptions: {
                    exportType: "named"
                }
            })),
            compressImage && pluginImageCompress(compressImage(["jpeg", "png", "ico", "svg"])),
            ...plugins
        ].filter(Boolean),
        tools: {
            rspack: {
                optimization: getOptimizationConfig(optimize),
                infrastructureLogging: verbose ? {
                    appendOnly: true,
                    level: "verbose",
                    debug: /PackFileCache/
                } : undefined
            }
        }
    };

    const transformedConfig = applyTransformers(config, transformers, {
        environment: "build",
        verbose
    });

    return defineConfig(transformedConfig);
}
