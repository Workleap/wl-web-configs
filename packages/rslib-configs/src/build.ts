import type { DistPathConfig, RsbuildPlugins, SourceMap } from "@rsbuild/core";
import { pluginReact, type PluginReactOptions } from "@rsbuild/plugin-react";
import { pluginSvgr, type PluginSvgrOptions } from "@rsbuild/plugin-svgr";
import { defineConfig, type Dts, type RsbuildConfigEntry, type RsbuildConfigOutputTarget, type Syntax } from "@rslib/core";
import { applyTransformers, RslibConfigTransformer } from "./applyTransformers.ts";

export type DefineBuildDefineReactPluginConfigFunction = (defaultOptions: PluginReactOptions) => PluginReactOptions;
export type DefineBuildSvgrPluginConfigFunction = (defaultOptions: PluginSvgrOptions) => PluginSvgrOptions;

export interface DefineBuildConfigOptions {
    entry?: RsbuildConfigEntry;
    format?: "esm" | "cjs";
    syntax?: Syntax;
    bundle?: boolean;
    tsconfigPath?: string;
    dts?: Dts;
    target?: RsbuildConfigOutputTarget;
    distPath?: DistPathConfig;
    plugins?: RsbuildPlugins;
    sourceMap?: boolean | SourceMap;
    react?: false | DefineBuildDefineReactPluginConfigFunction;
    svgr? : false | DefineBuildSvgrPluginConfigFunction;
    transformers?: RslibConfigTransformer[];
}

function defaultDefineReactPluginConfig(options: PluginReactOptions) {
    return options;
}

function defineSvgrPluginConfig(options: PluginSvgrOptions) {
    return options;
}

export function defineBuildConfig(options: DefineBuildConfigOptions = {}) {
    const {
        entry: entryValue,
        format = "esm",
        syntax = "esnext",
        bundle = false,
        tsconfigPath,
        dts = true,
        target = "web",
        distPath = {
            root: "./dist"
        },
        plugins = [],
        sourceMap = {
            js: "source-map",
            css: true
        },
        react = defaultDefineReactPluginConfig,
        svgr = defineSvgrPluginConfig,
        transformers = []
    } = options;

    if (!bundle && !tsconfigPath) {
        throw new Error("[rslib-configs] When the \"bundle\" option is \"false\", a \"tsconfigPath\" option must be provided.");
    }

    let entry = entryValue;

    if (!entry) {
        entry = {
            index: bundle ? ["./src/index.ts", "./src/index.js"] : "./src/**"
        };
    }

    const config = defineConfig({
        mode: "production",
        lib: [{
            format,
            syntax,
            bundle,
            dts
        }],
        source: {
            entry,
            tsconfigPath
        },
        output: {
            target,
            distPath,
            cleanDistPath: true,
            minify: false,
            sourceMap
        },
        plugins: [
            react && pluginReact(react({
                fastRefresh: false
            })),
            svgr && pluginSvgr(svgr({
                svgrOptions: {
                    exportType: "named"
                }
            })),
            ...plugins
        ].filter(Boolean),
    });

    const transformedConfig = applyTransformers(config, transformers, {
        environment: "build"
    });

    return defineConfig(transformedConfig);
}
