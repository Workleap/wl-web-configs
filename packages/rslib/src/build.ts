import { defineConfig, type EcmaScriptVersion } from "@rslib/core";

export interface DefineBuildConfigOptions {
    format: "esm" | "cjs";
    syntax: EcmaScriptVersion;
    bundle: boolean;
    dts: boolean;
}

export function defineBuildConfig(options: DefineBuildConfigOptions = {}) {
    return defineConfig({
        lib: [{
            format: "esm",
            syntax: "esnext",
            bundle: false,
            dts: true
        }],
        source: {
            entry: {
                index: "./src/**"
            },
            tsconfigPath: "./tsconfig.build.json"
        },
        output: {
            target: "web",
            distPath: {
                root: "./dist"
            },
            cleanDistPath: true,
            minify: false
        }
        // plugins: [pluginReact()]
    });
}

/*

- React
- SVGR
- CSS / CSS Modules
- Source maps?
- optimize

*/
