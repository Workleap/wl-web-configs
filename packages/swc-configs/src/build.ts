import type { Config } from "@swc/core";
import { applyTransformers, type SwcConfigTransformer } from "./applyTransformers";

export interface DefineBuildConfigOptions {
    // Any is also used for SWC "targets" type.
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    browsers: any;
    parser?: "ecmascript" | "typescript";
    transformers?: SwcConfigTransformer[];
}

export function defineBuildConfig(options: DefineBuildConfigOptions) {
    const {
        browsers,
        parser = "typescript",
        transformers = []
    } = options;

    const config: Config = {
        jsc: {
            parser: parser === "ecmascript"
                ? {
                    syntax: "ecmascript",
                    jsx: true
                }
                : {
                    syntax: "typescript",
                    tsx: true
                },
            // The output environment that the code will be compiled for.
            target: "es2022",
            // View https://swc.rs/docs/configuration/minification for options.
            minify: {
                compress: true,
                mangle: true
            },
            transform: {
                react: {
                    // Use "react/jsx-runtime".
                    runtime: "automatic",
                    // Use the native "Object.assign()" instead of "_extends".
                    useBuiltins: true
                }
            },
            // Import shims from an external module rather than inlining them in bundle files to greatly reduce the bundles size.
            // Requires to add "@swc/helpers" as a project dependency.
            externalHelpers: true
        },
        module: {
            // The output module resolution system that the code will be compiled for.
            type: "es6",
            // Prevent SWC from exporting the `__esModule` property.
            strict: true,
            // Preserve dynamic imports.
            ignoreDynamic: true
        },
        env: {
            targets: browsers
        }
    };

    const transformedConfig = applyTransformers(config, transformers, {
        environment: "build"
    });

    return transformedConfig;
}
