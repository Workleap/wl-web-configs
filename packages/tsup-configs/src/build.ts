import type { Options } from "tsup";
import { defineConfig, type DefineConfigOptions } from "./defineConfig.ts";

export const DefaultBuildOptions = {
    clean: true,
    dts: true,
    minify: true,
    splitting: false,
    treeshake: true,
    entry: ["./src"],
    outDir: "./dist",
    format: ["esm"],
    target: "esnext",
    platform: "browser"
};

export function defineBuildConfig(options?: DefineConfigOptions) {
    return defineConfig(DefaultBuildOptions as Options, options);
}

