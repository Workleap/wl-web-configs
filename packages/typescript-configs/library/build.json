{
    "$schema": "https://json.schemastore.org/tsconfig",
    "extends": "./lint.json",
    "compilerOptions": {
        "declaration": true,
        "noEmit": false,
        "sourceMap": true
    }
}
