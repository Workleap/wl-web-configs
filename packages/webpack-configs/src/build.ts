import type { Config as SwcConfig } from "@swc/core";
import HtmlWebpackPlugin from "html-webpack-plugin";
import MiniCssExtractPlugin from "mini-css-extract-plugin";
import { createRequire } from "node:module";
import path from "path";
import TerserPlugin from "terser-webpack-plugin";
import type { Configuration as WebpackConfig } from "webpack";
import { applyTransformers, type WebpackConfigTransformer } from "./transformers/applyTransformers.ts";

// Using node:module.createRequire until
// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import.meta/resolve
// is available
const require = createRequire(import.meta.url);

type MiniCssExtractPluginOptions = NonNullable<ConstructorParameters<typeof MiniCssExtractPlugin>[0]>;

export function defineBuildHtmlWebpackPluginConfig(options: HtmlWebpackPlugin.Options = {}) {
    const {
        template = "./public/index.html",
        ...rest
    } = options;

    return {
        ...rest,
        template
    } as HtmlWebpackPlugin.Options;
}

export function defineMiniCssExtractPluginConfig(options: MiniCssExtractPluginOptions = {}) {
    const {
        filename = "[name].[fullhash].css",
        ...rest
    } = options;

    return {
        ...rest,
        filename
    } as MiniCssExtractPluginOptions;
}

export interface DefineBuildConfigOptions {
    entry?: string;
    outputPath?: string;
    publicPath?: string;
    moduleRules?: NonNullable<WebpackConfig["module"]>["rules"];
    plugins?: WebpackConfig["plugins"];
    htmlWebpackPlugin?: HtmlWebpackPlugin.Options;
    miniCssExtractPlugin?: MiniCssExtractPluginOptions;
    minify?: boolean;
    cssModules?: boolean;
    postcssConfigFilePath?: string;
    swcConfig?: SwcConfig;
    transformers?: WebpackConfigTransformer[];
}

export async function defineBuildConfig(options: DefineBuildConfigOptions = {}) {
    const {
        entry = "./src/index.tsx",
        outputPath = path.resolve("dist"),
        // The trailing / is very important, otherwise paths will not be resolved correctly.
        publicPath = "http://localhost:8080/",
        moduleRules = [],
        plugins = [],
        htmlWebpackPlugin = defineBuildHtmlWebpackPluginConfig(),
        miniCssExtractPlugin = defineMiniCssExtractPluginConfig(),
        minify = true,
        cssModules = false,
        postcssConfigFilePath,
        swcConfig,
        transformers = []
    } = options;

    const config: WebpackConfig = {
        mode: "production",
        target: "web",
        entry,
        output: {
            path: outputPath,
            filename: "[name].[fullhash].js",
            publicPath,
            clean: true
        },
        optimization: minify
            ? {
                minimize: true,
                minimizer: [
                    new TerserPlugin({
                        minify: TerserPlugin.swcMinify,
                        terserOptions: {
                            compress: true,
                            mangle: true
                        }
                    })
                ]
            }
            : undefined,
        module: {
            rules: [
                {
                    test: /\.(js|jsx|ts|tsx)/i,
                    exclude: /node_modules/,
                    loader: require.resolve("swc-loader"),
                    options: swcConfig
                },
                {
                    // https://stackoverflow.com/questions/69427025/programmatic-webpack-jest-esm-cant-resolve-module-without-js-file-exten
                    test: /\.js/i,
                    include: /node_modules/,
                    resolve: {
                        fullySpecified: false
                    }
                },
                {
                    test: /\.css/i,
                    use: [
                        { loader: MiniCssExtractPlugin.loader },
                        {
                            loader: require.resolve("css-loader"),
                            options: cssModules
                                ? {
                                    // Must match the number of loaders applied before this one.
                                    importLoaders: 1,
                                    modules: true
                                }
                                : undefined
                        },
                        {
                            loader: require.resolve("postcss-loader"),
                            options: postcssConfigFilePath
                                ? {
                                    postcssOptions: {
                                        config: postcssConfigFilePath
                                    }
                                }
                                : undefined
                        }
                    ]
                },
                {
                    test: /\.svg/i,
                    loader: require.resolve("@svgr/webpack")
                },
                {
                    test: /\.(png|jpe?g|gif)/i,
                    type: "asset/resource"
                },
                ...moduleRules
            ]
        },
        resolve: {
            extensions: [".js", ".jsx", ".ts", ".tsx", ".css"]
        },
        plugins: [
            new HtmlWebpackPlugin(htmlWebpackPlugin),
            new MiniCssExtractPlugin(miniCssExtractPlugin),
            ...plugins
        ]
    };

    const transformedConfig = applyTransformers(config, transformers, {
        env: "build"
    });

    return transformedConfig;
}
