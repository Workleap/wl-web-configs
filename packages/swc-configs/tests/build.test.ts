import { defineBuildConfig } from "../src/build.ts";

const Browsers = [
    "last 2 versions",
    "> 0.2%",
    "Firefox ESR",
    "not dead"
];

test("specified \"browsers\" are set as \"env.targets\"", () => {
    const result = defineBuildConfig({
        browsers: Browsers
    });

    expect(result).toMatchSnapshot();
});

test("when \"parser\" is \"ecmascript\", the configuration parser is ecmascript", () => {
    const result = defineBuildConfig({
        browsers: Browsers,
        parser: "ecmascript"
    });

    expect(result).toMatchSnapshot();
});

test("when \"parser\" is \"typescript\", the configuration parser is typescript", () => {
    const result = defineBuildConfig({
        browsers: Browsers,
        parser: "typescript"
    });

    expect(result).toMatchSnapshot();
});

test("when \"parser\" is not set, the configuration parser is typescript", () => {
    const result = defineBuildConfig({
        browsers: Browsers
    });

    expect(result).toMatchSnapshot();
});
