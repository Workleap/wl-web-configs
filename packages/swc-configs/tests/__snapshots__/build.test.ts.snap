// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`specified "browsers" are set as "env.targets" 1`] = `
{
  "env": {
    "targets": [
      "last 2 versions",
      "> 0.2%",
      "Firefox ESR",
      "not dead",
    ],
  },
  "jsc": {
    "externalHelpers": true,
    "minify": {
      "compress": true,
      "mangle": true,
    },
    "parser": {
      "syntax": "typescript",
      "tsx": true,
    },
    "target": "es2022",
    "transform": {
      "react": {
        "runtime": "automatic",
        "useBuiltins": true,
      },
    },
  },
  "module": {
    "ignoreDynamic": true,
    "strict": true,
    "type": "es6",
  },
}
`;

exports[`when "parser" is "ecmascript", the configuration parser is ecmascript 1`] = `
{
  "env": {
    "targets": [
      "last 2 versions",
      "> 0.2%",
      "Firefox ESR",
      "not dead",
    ],
  },
  "jsc": {
    "externalHelpers": true,
    "minify": {
      "compress": true,
      "mangle": true,
    },
    "parser": {
      "jsx": true,
      "syntax": "ecmascript",
    },
    "target": "es2022",
    "transform": {
      "react": {
        "runtime": "automatic",
        "useBuiltins": true,
      },
    },
  },
  "module": {
    "ignoreDynamic": true,
    "strict": true,
    "type": "es6",
  },
}
`;

exports[`when "parser" is "typescript", the configuration parser is typescript 1`] = `
{
  "env": {
    "targets": [
      "last 2 versions",
      "> 0.2%",
      "Firefox ESR",
      "not dead",
    ],
  },
  "jsc": {
    "externalHelpers": true,
    "minify": {
      "compress": true,
      "mangle": true,
    },
    "parser": {
      "syntax": "typescript",
      "tsx": true,
    },
    "target": "es2022",
    "transform": {
      "react": {
        "runtime": "automatic",
        "useBuiltins": true,
      },
    },
  },
  "module": {
    "ignoreDynamic": true,
    "strict": true,
    "type": "es6",
  },
}
`;

exports[`when "parser" is not set, the configuration parser is typescript 1`] = `
{
  "env": {
    "targets": [
      "last 2 versions",
      "> 0.2%",
      "Firefox ESR",
      "not dead",
    ],
  },
  "jsc": {
    "externalHelpers": true,
    "minify": {
      "compress": true,
      "mangle": true,
    },
    "parser": {
      "syntax": "typescript",
      "tsx": true,
    },
    "target": "es2022",
    "transform": {
      "react": {
        "runtime": "automatic",
        "useBuiltins": true,
      },
    },
  },
  "module": {
    "ignoreDynamic": true,
    "strict": true,
    "type": "es6",
  },
}
`;
